eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelConfigOptions.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2021 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.channel
+
+import scodec.bits.{BitVector, ByteVector}
+
+/**
+ * Created by t-bast on 24/06/2021.
+ */
+
+/**
+ * Internal configuration option impacting the channel's structure or behavior.
+ * This must be set when creating the channel and cannot be changed afterwards.
+ */
+trait ChannelConfigOption {
+  def supportBit: Int
+}
+
+case class ChannelConfigOptions(activated: Set[ChannelConfigOption]) {
+
+  def hasOption(option: ChannelConfigOption): Boolean = activated.contains(option)
+
+  def bytes: ByteVector = toByteVector
+
+  def toByteVector: ByteVector = {
+    val indices = activated.map(_.supportBit)
+    if (indices.isEmpty) {
+      ByteVector.empty
+    } else {
+      // NB: when converting from BitVector to ByteVector, scodec pads right instead of left, so we make sure we pad to bytes *before* setting bits.
+      var buffer = BitVector.fill(indices.max + 1)(high = false).bytes.bits
+      indices.foreach(i => buffer = buffer.set(i))
+      buffer.reverse.bytes
+    }
+  }
+
+}
+
+object ChannelConfigOptions {
+
+  def standard: ChannelConfigOptions = ChannelConfigOptions(activated = Set(FundingPubKeyBasedChannelKeyPath))
+
+  def apply(options: ChannelConfigOption*): ChannelConfigOptions = ChannelConfigOptions(Set.from(options))
+
+  def apply(bytes: ByteVector): ChannelConfigOptions = {
+    val activated: Set[ChannelConfigOption] = bytes.bits.toIndexedSeq.reverse.zipWithIndex.collect {
+      case (true, 0) => FundingPubKeyBasedChannelKeyPath
+    }.toSet
+    ChannelConfigOptions(activated)
+  }
```

Comment:
```suggestion
  def apply(bytes: ByteVector): ChannelConfigOptions = {
    val activated: Set[ChannelConfigOption] = bytes.bits.toIndexedSeq.reverse.zipWithIndex.collect {
      resolveOption(FundingPubKeyBasedChannelKeyPath)
    }.toSet
    ChannelConfigOptions(activated)
  }

  def resolveOption(option: ChannelConfigOption): PartialFunction[(Boolean, Int), ChannelConfigOption] = {
    case (true, bit) if bit == option.supportBit => option
  }
```
Then we could do:
```scala
  def apply(bytes: ByteVector): ChannelConfigOptions = {
    val activated: Set[ChannelConfigOption] = bytes.bits.toIndexedSeq.reverse.zipWithIndex.collect {
      resolveOption(Feature1) orElse
      resolveOption(Feature2) orElse
      resolveOption(Feature3)
    }.toSet
    ChannelConfigOptions(activated)
  }
```

![image](https://user-images.githubusercontent.com/2028222/124250508-3dadb800-db25-11eb-861e-7404022b8694.png)

Reply:
Well, I'm oscillating between "this is really over-engineered" and "I feel mysteriously attracted to it" :laughing:

Reply:
What would be really great would be something that fails at compile time if we haven't handled all the possible `ChannelConfigOption`s (to make sure we don't forget to add a line here when we add a new option) but I'm not sure it's possible since here we're doing the reverse mapping (int to config options)...

Reply:
Use a `map` instead of a `collect`?

Reply:
A `map` would fail at runtime, but wouldn't help at compile time...at least with the current way I'm checking it, but I'm thinking that I should "reverse" the patter matching, I'll try something.

Reply:
Gotcha, this is because the space for the `supportBit` is the whole range of integers right?

Reply:
Yes exactly, I'm struggling with reflection to at least have a unit test that checks we're exhaustive. Apparently Scala 3 makes this better with macros, but for now there's no good solution...

Reply:
Would this be the missing bit (pun intended):

```scala
trait ChannelConfigOption {
  // @formatter:off
  def supportBit: ChannelOptionBit
  def name: String
  // @formatter:on
}

sealed trait ChannelOptionBit { def bit: Int }
case object FundingPubKeyBasedChannelKeyPathBit extends ChannelOptionBit { override def bit = 0 }
```

Reply:
This would still require narrowing the space somewhere around `bytes.bits.toIndexedSeq.reverse.zipWithIndex` though.

Reply:
I added a reflection test in https://github.com/ACINQ/eclair/pull/1852/commits/297d81aa5a2472145f7c9fad5cbcac76e7c7f2fc to make up for this (and I fixed the similar test we had in `FeatureSpec` which was completely broken :see_no_evil: ).